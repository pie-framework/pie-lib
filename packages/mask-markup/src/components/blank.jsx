import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { renderMath } from '@pie-lib/math-rendering';
import debug from 'debug';
import { DropTarget } from '@pie-lib/drag';
import { withStyles } from '@material-ui/core/styles';
import Chip from '@material-ui/core/Chip';
import classnames from 'classnames';
const log = debug('pie-lib:mask-markup:blank');
export const DRAG_TYPE = 'MaskBlank';

const useStyles = withStyles(theme => ({
  content: {
    border: `solid 0px ${theme.palette.primary.main}`,
    minWidth: '200px',
    padding: theme.spacing.unit
  },
  chip: {
    minWidth: '90px'
  },
  correct: {
    border: 'solid 1px green'
  },
  incorrect: {
    border: 'solid 1px red'
  }
}));

export class BlankContent extends React.Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    disabled: PropTypes.bool,
    value: PropTypes.string,
    classes: PropTypes.object,
    isOver: PropTypes.bool,
    dragItem: PropTypes.object,
    correct: PropTypes.bool,
    onChange: PropTypes.func
  };

  componentDidUpdate() {
    renderMath(this.rootRef);
  }

  render() {
    const { id, disabled, value, classes, isOver, dragItem, correct, onChange } = this.props;
    const label = dragItem && isOver ? dragItem.label : value;

    return (
      <Chip
        ref={ref => {
          //eslint-disable-next-line
          this.rootRef = ReactDOM.findDOMNode(ref);
        }}
        component="span"
        label={<span dangerouslySetInnerHTML={{ __html: label }} />}
        className={classnames(
          classes.chip,
          classes[correct !== undefined ? (correct ? 'correct' : 'incorrect') : undefined]
        )}
        variant={disabled ? 'outlined' : undefined}
        onDelete={value && !disabled ? () => onChange(id, undefined) : undefined}
      />
    );
  }
}

const StyledBlankContent = useStyles(BlankContent);

const connectedBlankContent = useStyles(({ connectDropTarget, ...props }) => {
  const { classes, isOver } = props;

  return connectDropTarget(
    <span className={classnames(classes.content, isOver && classes.over)}>
      <StyledBlankContent {...props} />
    </span>
  );
});

const tileTarget = {
  drop(props, monitor) {
    const draggedItem = monitor.getItem();
    log('props.instanceId', props.instanceId, 'draggedItem.instanceId:', draggedItem.instanceId);
    props.onChange(props.id, draggedItem.value);
  },
  canDrop(props, monitor) {
    const draggedItem = monitor.getItem();

    return draggedItem.instanceId === props.instanceId;
  }
};

const DropTile = DropTarget(DRAG_TYPE, tileTarget, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOver: monitor.isOver(),
  dragItem: monitor.getItem()
}))(connectedBlankContent);

// export default () => <div>hi</div>;
export default DropTile;
