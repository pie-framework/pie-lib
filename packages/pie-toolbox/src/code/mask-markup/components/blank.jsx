import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { renderMath } from '../../math-rendering-accessible';
import debug from 'debug';
import { DragSource, DropTarget } from '../../drag';
import { withStyles } from '@material-ui/core/styles';
import Chip from '@material-ui/core/Chip';
import classnames from 'classnames';
import { color } from '../../render-ui';
const log = debug('pie-lib:mask-markup:blank');
export const DRAG_TYPE = 'MaskBlank';

const useStyles = withStyles(() => ({
  content: {
    border: `solid 0px ${color.primary()}`,
    minWidth: '200px',
    touchAction: 'none',
    overflow: 'hidden',
    whiteSpace: 'nowrap', // Prevent line wrapping
  },
  chip: {
    backgroundColor: color.background(),
    border: `2px dashed ${color.text()}`,
    color: color.text(),
    minWidth: '90px',
    fontSize: 'inherit',
    minHeight: '32px',
    height: 'auto',
    maxWidth: '374px',
    position: 'relative',
  },
  chipLabel: {
    whiteSpace: 'pre-wrap',
    // Added for touch devices, for image content.
    // This will prevent the context menu from appearing and not allowing other interactions with the image.
    // If interactions with the image in the token will be requested we should handle only the context Menu.
    pointerEvents: 'none',
    '& img': {
      display: 'block',
      padding: '2px 0',
    },
  },
  hidden: {
    color: 'transparent',
    opacity: 0,
  },
  dragged: {
    position: 'absolute',
    left: 16,
    maxWidth: '60px',
  },
  correct: {
    border: `solid 1px ${color.correct()}`,
  },
  incorrect: {
    border: `solid 1px ${color.incorrect()}`,
  },
  over: {
    whiteSpace: 'nowrap',
    overflow: 'hidden',
  },
}));

export class BlankContent extends React.Component {
  static propTypes = {
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    disabled: PropTypes.bool,
    duplicates: PropTypes.bool,
    choice: PropTypes.object,
    classes: PropTypes.object,
    isOver: PropTypes.bool,
    dragItem: PropTypes.object,
    correct: PropTypes.bool,
    onChange: PropTypes.func,
  };
  constructor() {
    super();
    this.state = {
      height: 0,
    };
  }

  componentDidUpdate(prevProps) {
    renderMath(this.rootRef);
    const { choice: currentChoice } = this.props;
    const { choice: prevChoice } = prevProps;

    if (JSON.stringify(currentChoice) !== JSON.stringify(prevChoice)) {
      if (!currentChoice) {
        this.setState({
          height: 0,
        });
        return;
      }
      setTimeout(() => {
        this.setState({
          height: this.spanRef.offsetHeight,
        });
      }, 300);
    }
  }

  addDraggableFalseAttributes(parent) {
    parent.childNodes.forEach((elem) => {
      if (elem instanceof Element || elem instanceof HTMLDocument) {
        elem.setAttribute('draggable', false);
      }
    });
  }

  render() {
    const { disabled, choice, classes, isOver, dragItem, correct } = this.props;
    const draggedLabel = dragItem && isOver && dragItem.choice.value;
    const label = choice && choice.value;

    return (
      // TODO the Chip element is causing drag problems on touch devices. Avoid using Chip and consider refactoring the code. Keep in mind that Chip is a span with a button role, which interferes with seamless touch device dragging.
      <Chip
        clickable={false}
        disabled={true}
        ref={(ref) => {
          //eslint-disable-next-line
          this.rootRef = ReactDOM.findDOMNode(ref);
        }}
        component="span"
        label={
          <React.Fragment>
            <span
              className={classnames(classes.chipLabel, isOver && classes.over, {
                [classes.hidden]: draggedLabel,
              })}
              ref={(ref) => {
                if (ref) {
                  //eslint-disable-next-line
                  this.spanRef = ReactDOM.findDOMNode(ref);
                  ref.innerHTML = label || '';
                  this.addDraggableFalseAttributes(ref);
                }
              }}
            >
              {' '}
            </span>
            {draggedLabel && (
              <span
                className={classnames(classes.chipLabel, isOver && classes.over, classes.dragged)}
                ref={(ref) => {
                  if (ref) {
                    //eslint-disable-next-line
                    this.spanRef = ReactDOM.findDOMNode(ref);
                    ref.innerHTML = draggedLabel || '';
                    this.addDraggableFalseAttributes(ref);
                  }
                }}
              >
                {' '}
              </span>
            )}
          </React.Fragment>
        }
        className={classnames(classes.chip, isOver && classes.over, {
          [classes.correct]: correct !== undefined && correct,
          [classes.incorrect]: correct !== undefined && !correct,
        })}
        variant={disabled ? 'outlined' : undefined}
        style={{
          ...(this.state.height ? { height: this.state.height } : {}),
        }}
        classes={{
          label: isOver && classes.over,
        }}
      />
    );
  }
}
const StyledBlankContent = useStyles(BlankContent);

const connectedBlankContent = useStyles(({ connectDragSource, connectDropTarget, ...props }) => {
  const { classes, isOver } = props;

  return connectDropTarget(
    connectDragSource(
      <span className={classnames(classes.content, isOver && classes.over)}>
        <StyledBlankContent {...props} />
      </span>,
    ),
  );
});

const tileTarget = {
  drop(props, monitor) {
    const draggedItem = monitor.getItem();

    log('props.instanceId', props.instanceId, 'draggedItem.instanceId:', draggedItem.instanceId);

    if (draggedItem.id !== props.id) {
      props.onChange(props.id, draggedItem.choice.id);
    }

    return {
      dropped: draggedItem.id !== props.id,
    };
  },
  canDrop(props, monitor) {
    const draggedItem = monitor.getItem();

    return draggedItem.instanceId === props.instanceId;
  },
};

const DropTile = DropTarget(DRAG_TYPE, tileTarget, (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOver: monitor.isOver(),
  dragItem: monitor.getItem(),
}))(connectedBlankContent);

const tileSource = {
  canDrag(props) {
    return !props.disabled && !!props.choice;
  },
  beginDrag(props) {
    return {
      id: props.id,
      choice: props.choice,
      instanceId: props.instanceId,
      fromChoice: true,
    };
  },
  endDrag(props, monitor) {
    // this will be null if it did not drop
    const dropResult = monitor.getDropResult();

    if (!dropResult || (dropResult.dropped && !props.duplicates)) {
      const draggedItem = monitor.getItem();

      if (draggedItem.fromChoice) {
        props.onChange(props.id, undefined);
      }
    }
  },
};

const DragDropTile = DragSource(DRAG_TYPE, tileSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging(),
}))(DropTile);

export default DragDropTile;
