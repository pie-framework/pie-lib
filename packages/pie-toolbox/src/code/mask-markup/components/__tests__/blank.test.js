import * as React from 'react';
import { shallow } from 'enzyme';
import { BlankContent as Blank } from '../blank';

describe('Blank', () => {
  const onChange = jest.fn();
  const defaultProps = {
    disabled: false,
    value: 'Cow',
    classes: {},
    isOver: false,
    dragItem: {},
    correct: false,
    onChange,
  };
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<Blank {...defaultProps} />);
    instance = wrapper.instance();
  });

  describe('render', () => {
    it('renders correctly with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with disabled prop as true', () => {
      wrapper.setProps({ disabled: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with draggedItem', () => {
      wrapper.setProps({ dragItem: { choice: { value: 'Dog' } } });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with draggedItem and isOver', () => {
      wrapper.setProps({ dragItem: { choice: { value: 'Dog' } }, isOver: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('onDelete', () => {
    it('should be undefined if disabled is true', () => {
      wrapper.setProps({ disabled: true });

      expect(wrapper.props().onDelete).toEqual(undefined);
    });

    it('should be undefined if no value is set', () => {
      wrapper.setProps({ disabled: false, value: undefined });

      expect(wrapper.props().onDelete).toEqual(undefined);
    });
  });

  describe('updateDimensions', () => {
    let span;
    let rootRef;

    const setSpanDimensions = (height, width) => {
      Object.defineProperty(span, 'offsetHeight', { value: height, configurable: true });
      Object.defineProperty(span, 'offsetWidth', { value: width, configurable: true });
    };

    beforeEach(() => {
      wrapper = shallow(<Blank {...defaultProps} />);
      instance = wrapper.instance();

      span = document.createElement('span');
      rootRef = document.createElement('span'); // rootRef should be a span or div in the real component

      instance.spanRef = span;
      instance.rootRef = rootRef; // Attach rootRef

      Object.defineProperty(span, 'offsetHeight', { value: 0, configurable: true });
      Object.defineProperty(span, 'offsetWidth', { value: 0, configurable: true });
    });

    it('should update dimensions if span size exceeds the response area size', () => {
      setSpanDimensions(50, 50);

      instance.updateDimensions();

      expect(instance.state).toEqual({
        width: 74,
        height: 50,
      });
    });

    it('should not update dimensions if span size does not exceed the response area size', () => {
      wrapper.setProps({
        emptyResponseAreaHeight: 50,
        emptyResponseAreaWidth: 50,
      });
      setSpanDimensions(30, 30);

      instance.updateDimensions();

      expect(instance.state).toEqual({
        width: 0,
        height: 0,
      });
    });

    it('should handle non-numeric emptyResponseAreaHeight and emptyResponseAreaWidth', () => {
      wrapper.setProps({
        emptyResponseAreaHeight: 'non-numeric',
        emptyResponseAreaWidth: 'non-numeric',
      });
      setSpanDimensions(50, 50);

      instance.updateDimensions();

      expect(instance.state).toEqual({
        width: 74,
        height: 50,
      });
    });
  });

  describe('getRootDimensions', () => {
    it('should return state dimensions if set', () => {
      instance.setState({ height: 50, width: 50 });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 50,
        width: 50,
        minWidth: 90,
        minHeight: 32,
      });
    });

    it('should return state height and props width if state width is not set', () => {
      instance.setState({ height: 50, width: 0 });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 50,
        width: 0,
        minWidth: 90,
        minHeight: 32,
      });
    });

    it('should return props height and state width if state height is not set', () => {
      instance.setState({ height: 0, width: 50 });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 0,
        width: 50,
        minWidth: 90,
        minHeight: 32,
      });
    });

    it('should return props dimensions if state dimensions are zero', () => {
      instance.setState({ height: 0, width: 0 });
      wrapper.setProps({ emptyResponseAreaHeight: 60, emptyResponseAreaWidth: 60 });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 60,
        width: 60,
      });
    });

    it('should return state dimensions over props dimensions if both are set', () => {
      instance.setState({ height: 50, width: 50 });
      wrapper.setProps({ emptyResponseAreaHeight: 60, emptyResponseAreaWidth: 60 });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 50,
        width: 50,
      });
    });

    it('should return minWidth and minHeight if state and props dimensions are zero or undefined', () => {
      instance.setState({ height: 0, width: 0 });
      wrapper.setProps({ emptyResponseAreaHeight: undefined, emptyResponseAreaWidth: undefined });

      const dimensions = instance.getRootDimensions();

      expect(dimensions).toEqual({
        height: 0,
        width: 0,
        minWidth: 90,
        minHeight: 32,
      });
    });
  });
});
