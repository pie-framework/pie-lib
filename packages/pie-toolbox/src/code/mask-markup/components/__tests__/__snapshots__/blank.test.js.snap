// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blank render renders correctly with default props 1`] = `
<WithStyles(Chip)
  className="undefined"
  classes={
    Object {
      "label": false,
    }
  }
  clickable={false}
  component="span"
  disabled={true}
  label={
    <React.Fragment>
      <span
        className=""
      >
         
      </span>
    </React.Fragment>
  }
  style={Object {}}
/>
`;

exports[`Blank render renders correctly with disabled prop as true 1`] = `
<WithStyles(Chip)
  className="undefined"
  classes={
    Object {
      "label": false,
    }
  }
  clickable={false}
  component="span"
  disabled={true}
  label={
    <React.Fragment>
      <span
        className=""
      >
         
      </span>
    </React.Fragment>
  }
  style={Object {}}
  variant="outlined"
/>
`;

exports[`Blank render renders correctly with draggedItem 1`] = `
<WithStyles(Chip)
  className="undefined"
  classes={
    Object {
      "label": false,
    }
  }
  clickable={false}
  component="span"
  disabled={true}
  label={
    <React.Fragment>
      <span
        className=""
      >
         
      </span>
    </React.Fragment>
  }
  style={Object {}}
/>
`;

exports[`Blank render renders correctly with draggedItem and isOver 1`] = `
<WithStyles(Chip)
  className="undefined"
  classes={
    Object {
      "label": undefined,
    }
  }
  clickable={false}
  component="span"
  disabled={true}
  label={
    <React.Fragment>
      <span
        className="undefined"
      >
         
      </span>
      <span
        className=""
      >
         
      </span>
    </React.Fragment>
  }
  style={Object {}}
/>
`;
